### PR TITLE
datum/client_colour refactor.

### PR DESCRIPTION
**Pull request by Ghommie**
## About The Pull Request
Refactors datum/client_colour to allow client.color to be obtained from multiple `client_colour` datums instead of only the topmost one.

## Why It's Good For The Game
A more dynamic client_colour system. Also necessary for a future druggy status effect rework once the feature freeze is over.
Tested and working.

## Changelog
Not required.

### CODE REVIEW

File: code/modules/client/client_colour.dm [hunk 1]
```diff
@@ -1,72 +1,156 @@
...
+	  * We prioritize colours with higher priority (lower numbers), so they don't get overriden by less important ones:
+	  * eg: "Bloody screen" > "goggles colour" as the former is much more important
+	  */
+	var/priority = PRIORITY_NORMAL
+	///Will this client_colour prevent ones of lower priority from being applied?
+	var/override = FALSE
+	///IF non-zero, 'animate_client_colour(fade_in)' will be called instead of 'update_client_colour' when added.
+	var/fade_in = 0
+	///Same as above, but on removal.
+	var/fade_out = FALSE
```
Comment (MrPerson) [inline code/modules/client/client_colour.dm:33]: This isn't a bool

File: code/__HELPERS/matrices.dm [hunk 2]
```diff
@@ -193,3 +193,39 @@ round(cos_inv_third+sqrt3_sin, 0.001), round(cos_inv_third-sqrt3_sin, 0.001), ro
 		for(x in 1 to 4)
 			output[offset+x] = round(A[offset+1]*B[x] + A[offset+2]*B[x+4] + A[offset+3]*B[x+8] + A[offset+4]*B[x+12]+(y==5?B[x+16]:0), 0.001)
 	return output
+
+///Converts RGB shorthands into RGBA matrices complete of constants rows (ergo a 20 keys list in byond).
+/proc/color2fullRGBAmatrix(color)
+	if(istext(color))
+		var/list/L = ReadRGB(color)
+		if(!L)
+			return color_matrix_identity() //invalid format
```
Comment (Jared-Fogle) [inline code/__HELPERS/matrices.dm:202]: Why silently fail instead of something like `CRASH`?

File: code/__HELPERS/matrices.dm [hunk 3]
```diff
@@ -193,3 +193,39 @@ round(cos_inv_third+sqrt3_sin, 0.001), round(cos_inv_third-sqrt3_sin, 0.001), ro
 		for(x in 1 to 4)
 			output[offset+x] = round(A[offset+1]*B[x] + A[offset+2]*B[x+4] + A[offset+3]*B[x+8] + A[offset+4]*B[x+12]+(y==5?B[x+16]:0), 0.001)
 	return output
+
+///Converts RGB shorthands into RGBA matrices complete of constants rows (ergo a 20 keys list in byond).
+/proc/color2fullRGBAmatrix(color)
```
Comment (Jared-Fogle) [inline code/__HELPERS/matrices.dm:198]:
```suggestion
/proc/color_to_full_rgba_matrix(color)
```

...for consistency with snake_case naming (I know we have a lot of dumb proc names like this).

File: code/game/objects/effects/mines.dm [hunk 4]
```diff
@@ -177,15 +176,18 @@
...
-	animate(victim.client,color = red_splash, time = 10, easing = SINE_EASING|EASE_OUT)
-	sleep(10)
-	animate(victim.client,color = old_color, time = duration)//, easing = SINE_EASING|EASE_OUT)
-	sleep(duration)
-	to_chat(victim, "<span class='notice'>Your bloodlust seeps back into the bog of your subconscious and you regain self control.</span>")
-	qdel(chainsaw)
-	victim.log_message("exited a blood frenzy", LOG_ATTACK)
-	qdel(src)
+	var/datum/client_colour/C = victim.add_client_colour(/datum/client_colour/bloodlust)
+	QDEL_IN(C, 11)
```
Comment (Jared-Fogle) [inline code/game/objects/effects/mines.dm:180]:
```suggestion
	var/datum/client_colour/color = victim.add_client_colour(/datum/client_colour/bloodlust)
	QDEL_IN(color, 11)
```

Better variable names.

File: code/modules/client/client_colour.dm [hunk 5]
```diff
@@ -1,72 +1,155 @@
...
-	client_colours |= CC
-	sortTim(client_colours, /proc/cmp_clientcolour_priority)
-	update_client_colour()
-
-
-/*
-	Removes an instance of colour_type from the mob's client_colours list
-	colour_type - a typepath (subtyped from /datum/client_colour)
-*/
+	var/datum/client_colour/CC = new colour_type(src)
```
Comment (Jared-Fogle) [inline code/modules/client/client_colour.dm:64]: Better variable names--avoid abbreviating.

File: code/modules/client/client_colour.dm [hunk 6]
```diff
@@ -1,72 +1,155 @@
...
+  * In the case of multiple colours, they'll be converted to RGBA matrices for compatibility,
+  * summed together, and then each element divided by the number of matrices. (except we do this with lists because byond)
+  * target is the target variable.
+  */
+#define MIX_CLIENT_COLOUR(target)\
+	var/_our_colour;\
+	var/_number_colours = 0;\
+	var/_pool_closed = FALSE;\
+	for(var/_c in client_colours){\
+		var/datum/client_colour/_CC = _c;\
```
Comment (Jared-Fogle) [inline code/modules/client/client_colour.dm:97]: Better variable names, etc.

**Review by Mothblocks — CHANGES_REQUESTED** [review]
Mostly pedantic concerns--structure is good.

**Pull request merged**